core/blockchain/impl/block_tree_impl.cpp

```diff
@@ -36,9 +36,56 @@ namespace kagome::blockchain {
 
   BlockTreeImpl::TreeNode::TreeNode(primitives::BlockHash hash,
                                     primitives::BlockNumber depth,
-                                    const std::shared_ptr<TreeNode> &parent,
+                                    consensus::NextEpochDescriptor &&curr_epoch,
+                                    consensus::EpochIndex epoch_number,
+                                    consensus::NextEpochDescriptor &&next_epoch,
                                     bool finalized)
-      : block_hash{hash}, depth{depth}, parent{parent}, finalized{finalized} {}
+      : block_hash{hash},
+        depth{depth},
+        epoch_number(epoch_number),
+        finalized{finalized} {
+    this->epoch =
+        std::make_shared<consensus::NextEpochDescriptor>(std::move(curr_epoch));
+    this->next_epoch = *epoch == next_epoch
+                           ? epoch
+                           : std::make_shared<consensus::NextEpochDescriptor>(
+                               std::move(next_epoch));
+  }
+
+  BlockTreeImpl::TreeNode::TreeNode(
+      primitives::BlockHash hash,
+      primitives::BlockNumber depth,
+      const std::shared_ptr<TreeNode> &parent,
+      consensus::EpochIndex epoch_number,
+      boost::optional<consensus::NextEpochDescriptor> next_epoch_opt,
+      bool finalized)
+      : block_hash{hash},
+        depth{depth},
+        parent{parent},
+        epoch_number(epoch_number),
+        finalized{finalized} {
+    BOOST_ASSERT(parent != nullptr or next_epoch_opt.has_value());
+    if (parent) {
+      epoch = epoch_number != parent->epoch_number ? parent->next_epoch
+                                                   : epoch = parent->epoch;
+      next_epoch = parent->next_epoch;
+    } else {
+      epoch = std::make_shared<consensus::NextEpochDescriptor>(
```

Comment:
Do I understand correctly that next_epoch_opt is not checked in release builds and will throw an exception if empty? Is this the desired behaviour?

Reply:
This constructor is used at adding new block into block tree. Descriptor is usually existing only in first block of epoch.